modules/mod-deep-learning/ModelCard.cpp

```diff
@@ -0,0 +1,371 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   ModelCard.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "ModelCard.h"
+
+#include <wx/string.h>
+#include <wx/datetime.h>
+#include <wx/file.h>
+#include <wx/log.h>
+
+#include <rapidjson/document.h>
+#include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/filewritestream.h>
+#include <rapidjson/writer.h>
+#include <rapidjson/filereadstream.h>
+#include <rapidjson/error/en.h>
+
+#include "CodeConversions.h"
+
+namespace validators 
+{
+
+   void validateExists(const std::string &key, const Doc& doc)
+   {
+      if(!doc.IsObject())
+         throw InvalidModelCardDocument(XO("The provided JSON document is not an object."), "");
+      if(!doc.HasMember(key.c_str()))
+      {
+         
+         throw InvalidModelCardDocument(
+            XO("JSON document missing field: %s").Format(wxString(key)), "");
+      }
+   }
+
+   void throwTypeError(const std::string &key, const char *type, const Doc& doc)
+   {
+      throw InvalidModelCardDocument(
+         XO("field: %s is not of type: %s").Format(wxString(key), wxString(type)), "");
+   }
+
+   std::string toLowerCase(const std::string &string)
+   {
+      return audacity::ToUTF8(wxString(string).Lower());
+   }
+
+   std::string tryGetString(const std::string &key, const Doc& doc)
+   {
+      validateExists(key, doc);
+      if(!doc[key.c_str()].IsString())
+         throwTypeError(key, "string", doc);
+
+      return doc[key.c_str()].GetString();
+   }
+
+   std::string tryGetString(const std::string &key, const Doc& doc, const std::string &defaultValue)
+   {
+      try
+      {
+         return tryGetString(key, doc);
+      }
+      catch (const InvalidModelCardDocument &e)
+      {
+         return defaultValue;
+      }
```

Comment:
The implementation of the non-throwing function might be a little bit cheaper, if it and the throwing function are both implemented in terms of a third, non-throwing function, visible only in this .cpp file, that returns a success code; rather than having to catch an exception whenever defaulting of the value applies.

Similar for other pairs of functions below.

Reply:
e27545cbe5c2905cc9555f14c840eb8d3d870e5d

Reply:
94de8a3bbcca7f5c348a4b98865eed91a3b0a8b3

Reply:
94de8a3bbcca7f5c348a4b98865eed91a3b0a8b3